include/bio/map_io/header.hpp

```diff
@@ -0,0 +1,583 @@
+     * * **DS:** Description.  UTF-8 encoding may be used.
+     * * **DT:** Date the run was produced (ISO8601 date or date/time).
+     * * **FO:** Flow order. The array of nucleotide bases that correspond to the nucleotides used for each flow of each
+     *           read. Multi-base flows are encoded in IUPAC format, and non-nucleotide flows by various other
+     *           characters. Format : `/\*\|[ACMGRSVTWYHKDBN]+/`
+     * * **KS:** The array of nucleotide bases that correspond to the key sequence of each read.
+     * * **LB:** Library.
+     * * **PG:** Programs used for processing the read group.
+     * * **PI:** Predicted median insert size.
+     * * **PL:** Platform/technology used to produce the reads. Valid values : CAPILLARY, LS454, ILLUMINA, SOLID,
+     *           HELICOS, IONTORRENT, ONT, and PACBIO.
+     * * **PM:** Platform model. Free-form text providing further details of the platform/technology used.
+     * * **PU:** Platform unit (e.g. flowcell-barcode.lane for Illumina or slide for SOLiD). Unique identifier.
+     * * **SM:** Sample. Use pool name where a pool is being sequenced.
+     */
+    std::vector<std::pair<std::string, std::string>> read_groups{};
+    //!\}
+
+    /*!\name [PG] Programm information
+     * \brief You can directly edit these member variables.
+     * \{
+     */
+    //!\brief Stores information of the program/tool that was used to create the file.
+    struct program_info_t
+    {
+        std::string id{};                //!< A unique (file scope) id.
+        std::string name{};              //!< The official name.
+        std::string command_line_call{}; //!< The command line call that produces the file.
+        std::string previous{};          //!< The id of the previous program if program calls were chained.
+        std::string description{};       //!< A description of the program and/or program call.
+        std::string version{};           //!< The program/tool version.
+    };
+
+    std::vector<program_info_t> program_infos{}; //!< The list of program information.
+    //!\}
+
+    /*!\name [CO] Comments
+     * \brief You can directly edit these member variables.
+     * \{
+     */
+    std::vector<std::string> comments{}; //!< The list of comments.
+    //!\}
+};
+
+/*!\brief Reads the SAM header.
+ * \param[in] header_string  The full header as a std::string_view.
+ *
+ * \throws bio::map_io::format_error if any unexpected character or format is encountered.
+ *
+ * \details
+ *
+ * Reading the header format is done according to the official
+ * [SAM format specifications](https://samtools.github.io/hts-specs/SAMv1.pdf).
+ *
+ * The function throws a bio::map_io::format_error if any unknown tag was encountered. It will also fail if the format
+ * is not in a correct state (e.g. required fields are not given), but throwing might occur downstream of the actual
+ * error.
+ */
+void header::read(std::string_view header_string)
```

Comment:
This could be made private if the header is constructed from a full-header-string.